src/workerd/io/trace-common.h

```diff
@@ -220,30 +359,192 @@ struct Log final {
+  kj::Maybe<Info> info;
+  rpc::Trace::Span::SpanOutcome outcome;
+
+  void copyTo(rpc::Trace::SubrequestOutcome::Builder builder) const;
+  SubrequestOutcome clone() const;
+};
+
+struct SpanEvent final {
+  using Outcome = rpc::Trace::Span::SpanOutcome;
+  using Info = kj::OneOf<FetchResponseInfo, ActorFlushInfo, Tags>;
+  explicit SpanEvent(uint32_t id,
+      uint32_t parentId,
+      rpc::Trace::Span::SpanOutcome outcome,
+      bool transactional = false,
+      kj::Maybe<Info> maybeInfo = kj::none,
+      Tags tags = nullptr);
+  SpanEvent(rpc::Trace::Span::Reader reader);
+  SpanEvent(SpanEvent&&) = default;
+  SpanEvent& operator=(SpanEvent&&) = default;
+  KJ_DISALLOW_COPY(SpanEvent);
+
+  uint32_t id;
+  uint32_t parent;
+  rpc::Trace::Span::SpanOutcome outcome;
+  bool transactional;
+  kj::Maybe<Info> info;
+  Tags tags;
+
+  void copyTo(rpc::Trace::Span::Builder builder) const;
+  SpanEvent clone() const;
+};
+
+struct Mark final {
+  explicit Mark(kj::String name);
+  Mark(rpc::Trace::Mark::Reader reader);
+  Mark(Mark&&) = default;
+  Mark& operator=(Mark&&) = default;
+  KJ_DISALLOW_COPY(Mark);
 
-  void copyTo(rpc::Trace::Exception::Builder builder);
+  kj::String name;
+
+  void copyTo(rpc::Trace::Mark::Builder builder) const;
+  Mark clone() const;
 };
 
+struct Metric final {
+  using Key = kj::OneOf<kj::String, uint32_t>;
+  using Value = kj::OneOf<double, int64_t, uint64_t>;
+  using Type = rpc::Trace::Metric::Type;
+
+  explicit Metric(Type type, Key key, Value value);
+
+  template <IsEnum K>
+  explicit Metric(Type type, K key, Value value)
+      : Metric(type, static_cast<uint32_t>(key), kj::mv(value)) {}
+
+  Metric(rpc::Trace::Metric::Reader reader);
+  Metric(Metric&&) = default;
+  Metric& operator=(Metric&&) = default;
+  KJ_DISALLOW_COPY(Metric);
+
+  Type type;
+  Key key;
+  Value value;
+
+  bool keyMatches(kj::OneOf<kj::StringPtr, uint32_t> key);
+
+  template <IsEnum Key>
+  inline bool keyMatches(Key key) {
+    return keyMatches(static_cast<uint32_t>(key));
+  }
+
+  void copyTo(rpc::Trace::Metric::Builder builder) const;
+  Metric clone() const;
+};
+using Metrics = kj::Array<Metric>;
+
+struct Dropped final {
+  explicit Dropped(uint32_t start, uint32_t end);
+  Dropped(rpc::Trace::Dropped::Reader reader);
+  Dropped(Dropped&&) = default;
+  Dropped& operator=(Dropped&&) = default;
+  KJ_DISALLOW_COPY(Dropped);
+
+  uint32_t start;
+  uint32_t end;
+
+  void copyTo(rpc::Trace::Dropped::Builder builder) const;
+  Dropped clone() const;
+};
+
+using EventDetail = kj::OneOf<LogV2,
+    Exception,
+    DiagnosticChannelEvent,
+    Mark,
+    Metrics,
+    Subrequest,
+    SubrequestOutcome,
+    Tags>;
+
+// ======================================================================================
 // Represents a trace span. `Span` objects are delivered to `SpanObserver`s for recording. To
-// create a `Span`, use a `SpanBuilder`.
+// create a `Span`, use a `SpanBuilder`. (Used in the legacy trace api)
```

Comment:
Maybe we should rename it to LegacySpan?

Reply:
Intentionally not changing this in this PR.